sklearn/neighbors/_kde.py

```diff
@@ -216,28 +216,46 @@ def fit(self, X, y=None, sample_weight=None):
         """
         algorithm = self._choose_algorithm(self.algorithm, self.metric)
 
+        if sample_weight is not None:
+            sample_weight = _check_sample_weight(
+                sample_weight, X, dtype=np.float64, ensure_non_negative=True
+            )
+            normalized_sample_weight = sample_weight / sample_weight.sum()
+            n_effective_samples = 1 / np.sum(normalized_sample_weight**2)
```

Comment:
@Charlie-XIAO would you have a reference as to why you use (sorry I may just be missing something here): 
```n_effective_samples = 1 / np.sum(normalized_sample_weight**2)```

Reply:
Scipy doc of `gaussian_kde`: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.gaussian_kde.html. It has the `neff` thing which is short for *effective number* of samples, computed as

```
neff = sum(weights)^2 / sum(weights^2)
```

Reply:
This is breaking the usual "reweighing is repeating" semantics for `sample_weight` that we would like to enforce in all scikit-learn estimators.

```python
import matplotlib.pyplot as plt
from scipy.stats import gaussian_kde
import numpy as np

rng = np.random.default_rng(seed=42)
n_samples = 30
data = np.concatenate(
    [
        rng.standard_normal(size=n_samples // 2) - 1,
        3 * rng.standard_normal(size=n_samples // 2) + 5,
    ]
)
weights = rng.integers(low=0, high=5, size=n_samples)

bw_method = "scott"
bw_method = "silverman"
kde_weights = gaussian_kde(
    data, bw_method=bw_method, weights=weights.astype(np.float64)
)
kde_repetitions = gaussian_kde(np.repeat(data, weights), bw_method=bw_method)

x = np.linspace(data.min(), data.max(), 100)
plt.plot(x, kde_weights.evaluate(x), label="weights")
plt.plot(x, kde_repetitions.evaluate(x), label="repetitions")
_ = plt.legend()
```

![image](https://github.com/user-attachments/assets/7ba6b8d3-d726-4855-b075-230b950f2f4e)

Reply:
Note that in scipy's Gaussian KDE implementation, if you use a fixed bandwidth (e.g. `bw_method = 0.3` in the code above), then the repetitions vs weights equivalence seems to be restored:

![image](https://github.com/user-attachments/assets/7ab8ea42-4dca-4c00-be4d-d95eabecc8ed)

Reply:
Scipy does not specify `ddof=0`, and uses `neff = sum(weights)^2 / sum(weights^2)` (for scott and silverman), which is essentially the same as currently in this PR, so in https://github.com/scikit-learn/scikit-learn/pull/27971#discussion_r1763244035 I think the small deviation comes from `ddof`?

Regarding scott and silverman, I briefly searched through the references in the scipy docs, did not find something useful in the weighted case so I don't know where the scipy formula really comes from...

In https://github.com/Charlie-XIAO/scikit-learn/pull/3/files#r1761412299 you changed to `neff = sum(weights)`, I wonder what's the intuition behind that 🤔

Reply:
I tried patching scipy's implementation and passing `bias=True` / `ddof=0` to its own internal call to `np.cov` fixes the remaining discrepancy (as it does for scikit-learn's implementation).